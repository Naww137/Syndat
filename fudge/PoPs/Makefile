# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

SUB_DIRS = quantities chemicalElements families decays atomic Test

.PHONY: clean realclean doTarget

default:

clean:
	rm -f *.pyc
	$(MAKE) doTarget MODE=clean

realclean: clean
	$(MAKE) doTarget MODE=realclean

doTarget:
	SAVED_PWD=`pwd`; \
	for directory in $(SUB_DIRS); do cd $$directory; $(MAKE) $(MODE); cd $$SAVED_PWD; done
