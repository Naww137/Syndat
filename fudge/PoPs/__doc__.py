# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This is the 'Properties of Particles' or PoPs package. It defines a set of Python classes
for storing intrinsic particle properties like charge, mass, spin, etc.
"""
