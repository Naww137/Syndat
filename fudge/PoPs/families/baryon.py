# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the baryon class definitions.
Baryons include the neutron and proton.
"""

from . import particle as particleModule

class Alias( particleModule.Alias ) :

    moniker = 'baryonAlias'

class Particle( particleModule.Particle ) :

    moniker = 'baryon'
    familyOrder = 2
    alias = Alias

    def __lt__( self, other ) :

        if( self.familyOrderLessThan( other ) ) : return( True )        # Also checks isinstance of other.
        if( self.familyOrder != other.familyOrder ) : return( False )
        return( self.getMass( 'amu' ) < other.getMass( 'amu' ) )

class Suite( particleModule.Suite ) :

    moniker = 'baryons'
    particle = Particle
