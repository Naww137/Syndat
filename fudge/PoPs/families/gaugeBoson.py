# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the gaugeBoson class definitions.
Gauge bosons include the photon, gluon, W and Z particles
"""

from . import particle as particleModule

class Alias( particleModule.Alias ) :

    moniker = 'gaugeBosonAlias'

class Particle( particleModule.Particle ) :

    moniker = 'gaugeBoson'
    familyOrder = 0
    alias = Alias

    def __lt__( self, other ) :

        if( self.familyOrderLessThan( other ) ) : return( True )        # Also checks isinstance of other.
        if( self.familyOrder != other.familyOrder ) : return( False )
        return( False )                                                 # Currently, only supports photon.

class Suite( particleModule.Suite ) :

    moniker = 'gaugeBosons'
    particle = Particle
