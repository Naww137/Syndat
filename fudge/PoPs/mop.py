# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

__doc__ = """
This module contains classes to handle a MoP (Map of PoPs) instance.
"""

import os
import re

from LUPY import ancestry as ancestryModule
from . import database as databaseModule

class FormatVersion :

    version_0_1 = '0.1'

    allowed = ( version_0_1, )

class Mop(ancestryModule.AncestryIO):

    moniker = 'mop'

    def __init__(self, name, path = './', format=FormatVersion.version_0_1):

        """
        Constructor for a Mop instance.

        :param name:            The name of the library.
        :param path:            The path to the mop file if read in from disk.
        :param format:          The format version for the mop.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( name, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__name = name 

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__path = path

        if( format not in FormatVersion.allowed ) : raise ValueError( 'Invalid format "%s."' % format )
        self.__format = format

        self.__entries = []
        self.__pops = databaseModule.Database( 'Internal', '0.1' )

    def __len__( self ) :
        """Returns the number of entries in self."""

        return( len( self__entries ) )

    def __getitem__( self, index ) :
        """
        Returns the (index-1)^th item of self.

        :param index:       The index of the iter to return.
        """

        return( self.__entries[index] )

    @property
    def format( self ) :
        """Returns the format self is stored as."""

        return( self.__format )

    @property
    def name( self ) :
        """Returns self's name."""

        return( self.__name )

    @property
    def path( self ) :
        """Returns the path to self. If the data in self was read from disk, this is the path to the file."""

        return( self.__path )

    @property
    def fileName( self ) :
        """Same the path method."""

        return( self.__path )

    def append( self, entry ) :
        """Appends *entry* to the end of self's list of entries."""

        if( not( isinstance( entry, EntryBase ) ) ) : raise TypeError( 'Invalid instance to add.' )

        self.__entries.append( entry )
        entry.setAncestor( self )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns a list of items. Each item contains information about a PoPs database where one or more matches to *re_pid* exists. 
        Each item contains the list [ ids, all_ids, fileName, pops ]) where:

            -) ids:         The list of PoPs ids in a PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The file path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]).  *re_pid* can be a str instance or a *re.compile* instance.
        Note, even if *readPoPs* is False, the returned pops will not be None if it had previously been read in.
        
        :param re_pid:          A str or a *re.compile* pattern representing the ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        matches = []
        for entry in self.__entries : matches += entry.fetch( re_pid, readPoPs )

        return( matches )

    def toXML_strList( self, indent = "", **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        indent2 = indent + kwargs.get( 'incrementalIndent', '  ' )

        format = kwargs.get( 'format', FormatVersion.version_0_1 )
        if( format not in FormatVersion.allowed ) : raise TypeError( 'Invalid format = "%s".' % format )

        XML_list = [ '%s<%s name="%s" format="%s">' % ( indent, self.moniker, self.name, format ) ]
        for entry in self.__entries : XML_list += entry.toXML_strList( indent2, **kwargs )
        XML_list[-1] +=  '</%s>' % self.moniker

        return( XML_list )

    @staticmethod
    def parseNode( node, path ) :
        """
        Creates a Mop instance from an XML mop element.

        :param node:        XML element to parse.

        :return:            Mop instance.
        """

        if( node.tag != Mop.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Mop.moniker, node.tag ) )

        format = node.get( 'format', None )
        if( format is None ) : raise ValueError( 'Mop node does not have a "format" attribute.' )

        name = node.get( 'name', None )
        if( name is None ) : raise ValueError( 'Mop node does not have "name" attribute.' )

        mop = Mop( name, path = path )

        settings = { 'format' : format }
        for child in node :
            if( child.tag == Import.moniker ) :
                mop.append( Import.parseNode( settings, child ) )
            elif( child.tag == PoPs.moniker ) :
                mop.append( PoPs.parseNode( settings, child ) )
            else :
                raise ValueError( 'Invalid child tag "%s" for mop file.' % child.tag )

        return( mop )

class EntryBase(ancestryModule.AncestryIO) :
    """
    Base class for mop entries.
    """

    def __init__( self, path ) :
        """
        Base constructor for mop entries.

        :param path:            Path to the file for the entry.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Path must be a string.' )
        self.__path = path

    @property
    def fileName( self ) :
        """Returns the absolute path to the file pointed to by the 'path' attribute."""

        filename = self.path
        if( filename != os.sep ) : filename = os.path.join( os.path.dirname( os.path.realpath( self.ancestor.fileName ) ), filename )
        return( filename )

    @property
    def path( self ) :
        """Returns the path of the mop file. This may be absolute or relative."""

        return( self.__path )

class PoPs(EntryBase):
    """
    Class for storing a GNDS mop's PoPs instance.
    """

    moniker = 'pops'

    def __init__( self, path, ids ) :
        """
        Constructor for a mop's PoPs instance.

        :param path:            Path to the GNDS PoPs.database file.
        :param ids:             The list of particles ids in the PoPs.database file.
        """

        EntryBase.__init__(self, path)
        self.__pops = None
        self.__ids = ids

    @property
    def ids( self ) :
        """Returns the list of ids."""

        return( self.__ids )

    @property
    def pops( self ) :
        """Returns a reference to the PoPs.database instance. If the PoPs file has not be read in (see method *readIfNeeded*), None is returned."""

        return( self.__pops )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns the list [ ids, all_ids, fileName, pops ] if matches to *re_pid* are found in *self* and an empty list otherwise. 

            -) ids:         The list of PoPs ids in the PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]). *re_pid* can be a str instance or a *re.compile* instance.
        
        :param re_pid:          A str or *re.compile* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element.
        """

        if( isinstance( re_pid, str ) ) : re_pid = re.compile( re_pid )
        matches = [ pid for pid in self.ids if re_pid.fullmatch( pid ) ]

        if( len( matches ) > 0 ) :
            matches = [ [ matches, self.ids, self.fileName, self.readIfNeeded( readPoPs ) ] ]

        return( matches )

    def readIfNeeded( self, readPoPs = True ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in by a prior call and *readPoPs* 
        is True (default). Also, returns the value of *self.__pops*. The returned value will only be *None* if *readPoPs* 
        is *False* and all prior calls to this method had "*readPoPs* = False".

        :param readPoPs:        If *True*, the pops file is read.
        :return:                The value if *self.__pops*.
        """

        if( ( self.__pops is None ) and readPoPs ) :
            self.__pops = databaseModule.read( self.fileName )
            self.__pops.setAncestor( self )

        return( self.__pops )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s">' % ( indent, self.moniker, self.path ) + ' '.join( self.__ids ) + '</%s>' % self.moniker ] )

    @classmethod
    def parseNodeUsingClass(settings, node):
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != PoPs.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( PoPs.moniker, node.tag ) )

        path = node.get( 'path', None )
        ids = node.text.split( )

        pops = PoPs( path, ids )

        return( pops )

class Import(EntryBase):
    """
    Class for storing a GNDS mop's Import instance.
    """

    moniker = 'import'

    def __init__( self, path ) :
        """
        Constructor for a mop's import instance.

        :param path:            Path to the GNDS mop file.
        """

        EntryBase.__init__(self, path)
        self.__mop = None

    @property
    def mop( self ) :
        """Returns a reference to self.__mop."""

        return( self.__mop )

    def fetch( self, re_pid, readPoPs ) :
        """
        See documentation for Mop.fetch.
        
        :param re_pid:          A str or *re* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        self.readIfNeeded( )
        return( self.__mop.fetch( re_pid, readPoPs ) )

    def readIfNeeded( self ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in.
        """

        if( self.__mop is None ) :
            self.__mop = Mop.readXML_file( self.fileName )
            self.__mop.setAncestor( self )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s"/>' % ( indent, self.moniker, self.path ) ] )

    @staticmethod
    def parseNode( settings, node ) :
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != Import.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Import.moniker, node.tag ) )

        path = node.get( 'path', None )

        _import = Import( path )

        return( _import )
