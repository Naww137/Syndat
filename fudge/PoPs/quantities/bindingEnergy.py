# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the binding energy classes.
Binding energy is typically stored in units of eV.
"""

from . import quantity as quantityModule

baseUnit = quantityModule.stringToPhysicalUnit( 'eV' )

class Double( quantityModule.Double ) :

    baseUnit = baseUnit

class Suite( quantityModule.NumberSuite ) :

    moniker = 'bindingEnergy'
    _allowedClasses = [ Double ]
