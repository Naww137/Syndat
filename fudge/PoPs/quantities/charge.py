# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the charge classes.
Charge is typically given in units of the electron charge 'e'.
"""

from . import quantity as quantityModule

baseUnit = quantityModule.stringToPhysicalUnit( 'e' )

class Integer( quantityModule.Integer ) :

    baseUnit = baseUnit

class Suite( quantityModule.NumberSuite ) :

    moniker = 'charge'
    _allowedClasses = [ Integer ]
