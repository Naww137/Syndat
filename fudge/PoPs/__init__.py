# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
Top-level __init__ for the PoPs package.
"""

POPS_MAJORVERSION = 0
POPS_MINORVERSION = 1
POPS_PATCHLEVEL = 0
__version__ = '%s.%s.%s' % ( POPS_MAJORVERSION, POPS_MINORVERSION, POPS_PATCHLEVEL )

from .__doc__ import __doc__
