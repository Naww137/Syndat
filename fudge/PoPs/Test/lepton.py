#! /usr/bin/env python3

# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

from PoPs.quantities import quantity as quantityModule
from PoPs.quantities import mass as massModule
from PoPs.quantities import spin as spinModule
from PoPs.quantities import parity as parityModule
from PoPs.quantities import charge as chargeModule
from PoPs.quantities import halflife as halflifeModule

from PoPs.families import lepton as leptonModule

electron = leptonModule.Particle( 'e-', generation = 'electronic' )

mass = massModule.Double( 'base', 5.48579909070e-4, quantityModule.stringToPhysicalUnit( 'amu' ) )
electron.mass.add( mass )

charge = chargeModule.Integer( 'base', -1, quantityModule.stringToPhysicalUnit( 'e' ) )
electron.charge.add( charge )

halflife = halflifeModule.String( 'base', 'stable', quantityModule.stringToPhysicalUnit( 's' ) )
electron.halflife.add( halflife )

spin = spinModule.Fraction( 'base', spinModule.Fraction.toValueType( '1/2' ), quantityModule.stringToPhysicalUnit( 'hbar' ) )
electron.spin.add( spin )

parity = parityModule.Integer( 'base', 1, quantityModule.stringToPhysicalUnit( '' ) )
electron.parity.add( parity )

print( electron.toXML( ) )
print()
electron2 = leptonModule.Particle.parseXMLString(electron.toXML())
if( electron.toXML( ) != electron2.toXML( ) ) : raise Exception( 'Fix me' )

suite = leptonModule.Suite( )
suite.add( electron )
print( suite.toXML( ) )
suite2 = leptonModule.Suite.parseXMLString( suite.toXML())
if( suite2.toXML( ) != suite.toXML( ) ) : raise Exception( 'Fix me' )
