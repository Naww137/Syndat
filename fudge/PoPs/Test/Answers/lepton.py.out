<lepton id="e-" generation="electronic">
  <mass>
    <double label="base" value="5.4857990907e-4" unit="amu"/></mass>
  <spin>
    <fraction label="base" value="1/2" unit="hbar"/></spin>
  <parity>
    <integer label="base" value="1"/></parity>
  <charge>
    <integer label="base" value="-1" unit="e"/></charge>
  <halflife>
    <string label="base" value="stable" unit="s"/></halflife></lepton>

<leptons>
  <lepton id="e-" generation="electronic">
    <mass>
      <double label="base" value="5.4857990907e-4" unit="amu"/></mass>
    <spin>
      <fraction label="base" value="1/2" unit="hbar"/></spin>
    <parity>
      <integer label="base" value="1"/></parity>
    <charge>
      <integer label="base" value="-1" unit="e"/></charge>
    <halflife>
      <string label="base" value="stable" unit="s"/></halflife></lepton></leptons>
